File: ctci/java/ch1-arrays-and-strings/q4_palindrome_permutation/Question.java
package q4_palindrome_permutation;

/**
 * Created by jjlee on 7/10/16.
 */
public class Question {
    /* get character value */
    public static int getCharValue(char c) {
        char lowerC = Character.toLowerCase(c);
        if (lowerC >= 'a' && lowerC <= 'z')
            return lowerC - 'a';
        return -1;
    }

    /* use a bit vector and assume alphabets where lowercase == uppercase (i.e. 26 characters) */
    public static boolean isPermutationOfPalindrome(String s) {
        int countsOdd = 0;
        for (int i = 0; i < s.length(); i++) {
            int shift = getCharValue(s.charAt(i));
            if (shift != -1) {
                int bitMask = 1 << shift;
                countsOdd ^= bitMask;
            }
        }
        return (countsOdd & (countsOdd - 1)) == 0;
    }
}

File: ctci/python/ch1-arrays-and-strings/q4-palindrome-permutation/Question.py
# get character value
def get_char_value(c):
    lower_c, a, z = ord(c.lower()), ord('a'), ord('z')
    return lower_c - a if a <= lower_c <= z else -1


# use a bit vector and assume alphabets where lowercase == uppercase (i.e. 26 characters)
def is_permutation_of_palindrome(s):
    counts_odd = 0
    for c in s:
        shift = get_char_value(c)
        if shift != -1:
            bit_mask = 1 << shift
            counts_odd ^= bit_mask
    return (counts_odd & (counts_odd - 1)) == 0